### PR TITLE
dotnet-ai-ml

### PR DESCRIPTION
**Pull request by jeffschwMSFT**
This is the updated PR for the AI SDK technology selection PR (sorry the previous PR/branch was lost in the private to public move)

### CODE REVIEW

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,350 @@
+---
+name: technology-selection
+description: "Guides technology selection and implementation of AI and ML features in .NET 8+ applications using ML.NET, Microsoft.Extensions.AI (MEAI), Microsoft Agent Framework (MAF), GitHub Copilot SDK, ONNX Runtime, and LLamaSharp. Covers the full spectrum from classic ML through modern LLM orchestration to local inference. Use when adding classification, regression, clustering, anomaly detection, recommendation, LLM integration (text generation, summarization, reasoning), RAG pipelines with vector search, agentic workflows with tool calling, Copilot extensions, or custom model inference via ONNX Runtime to a .NET project. DO NOT USE FOR projects targeting .NET Framework (requires .NET 8+), the task is pure data engineering or ETL with no ML/AI component, or the project needs a custom deep learning training loop (use Python with PyTorch/TensorFlow, then export to ONNX for .NET inference)."
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
description: "Guides technology selection and implementation of AI and ML features in .NET 8+ applications using ML.NET, Microsoft.Extensions.AI (MEAI), Microsoft Agent Framework (MAF), GitHub Copilot SDK, ONNX Runtime, and OllamaSharp. Covers the full spectrum from classic ML through modern LLM orchestration to local inference. Use when adding classification, regression, clustering, anomaly detection, recommendation, LLM integration (text generation, summarization, reasoning), RAG pipelines with vector search, agentic workflows with tool calling, Copilot extensions, or custom model inference via ONNX Runtime to a .NET project. DO NOT USE FOR projects targeting .NET Framework (requires .NET 8+), the task is pure data engineering or ETL with no ML/AI component, or the project needs a custom deep learning training loop (use Python with PyTorch/TensorFlow, then export to ONNX for .NET inference)."
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+| Natural language understanding, generation, summarization, reasoning over unstructured text (single prompt → response, no tool calling) | **LLM via Microsoft.Extensions.AI** (`IChatClient`) | Requires language model capabilities beyond pattern matching; no orchestration needed |
+| Agentic workflows: tool/function calling, multi-step reasoning, agent loops, multi-agent collaboration | **Microsoft Agent Framework** (`Microsoft.Agents.AI`) built on top of **Microsoft.Extensions.AI** | Requires orchestration, tool dispatch, iteration control, and guardrails that `IChatClient` alone does not provide |
+| Building GitHub Copilot extensions, custom agents, or developer workflow tools | **GitHub Copilot SDK** (`GitHub.Copilot.SDK`) | Integrates with the Copilot agent runtime for IDE and CLI extensibility |
+| Running a pre-trained or fine-tuned custom model in production | **ONNX Runtime** (`Microsoft.ML.OnnxRuntime`) | Hardware-accelerated inference, model-format agnostic |
+| Local/offline LLM inference with no cloud dependency | **LLamaSharp** with quantized GGUF models | Privacy-sensitive, air-gapped, or cost-constrained scenarios |
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
| Local/offline LLM inference with no cloud dependency | **OllamaSharp** with local [AI models supported by Ollama](https://ollama.com/search) | Privacy-sensitive, air-gapped, or cost-constrained scenarios |
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+<!-- Orchestration (agents, workflows, tools, memory) — prerelease; use dotnet add package Microsoft.Agents.AI --prerelease -->
+<PackageReference Include="Microsoft.Agents.AI" Version="1.*-*" />
+
+<!-- Cloud LLM provider (pick one) -->
+<PackageReference Include="Azure.AI.OpenAI" Version="2.*" />
+<!-- OR -->
+<PackageReference Include="OpenAI" Version="2.*" />
+
+<!-- Local LLM inference -->
+<PackageReference Include="LLamaSharp" Version="0.*" />
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
<PackageReference Include="OllamaSharp" Version="5.*" />
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+
+## Workflow
+
+### Step 1: Classify the task using the decision tree
+
+Evaluate the developer's task against this decision tree and select the appropriate technology. State which branch applies and why.
+
+| Task type | Technology | Rationale |
+|-----------|-----------|-----------|
+| Structured/tabular data: classification, regression, clustering, anomaly detection, recommendation | **ML.NET** (`Microsoft.ML`) | Deterministic, reproducible, no cloud dependency, purpose-built for these tasks |
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
| Structured/tabular data: classification, regression, clustering, anomaly detection, recommendation | **ML.NET** (`Microsoft.ML`) | Reproducible (given a fixed seed and dataset), no cloud dependency, purpose-built models for these tasks |
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+| Local/offline LLM inference with no cloud dependency | **LLamaSharp** with quantized GGUF models | Privacy-sensitive, air-gapped, or cost-constrained scenarios |
+| Semantic search, RAG, or embedding storage | **Microsoft.Extensions.VectorData.Abstractions** + a vector database provider (e.g., Azure AI Search, Milvus, MongoDB, pgvector, Pinecone, Qdrant, Redis, SQL) | Provider-agnostic abstractions for vector similarity search; pair with a database-specific connector package (many are moving to community toolkits) |
+| Ingesting, chunking, and loading documents into a vector store | **Microsoft.Extensions.AI.DataIngestion** (preview) + **Microsoft.Extensions.VectorData.Abstractions** (MEVD) | Handles document parsing, text chunking, embedding generation, and upserting into a vector database; pairs with Microsoft.Extensions.VectorData.Abstractions |
+| Both structured ML predictions AND natural language reasoning | **Hybrid**: ML.NET for predictions + LLM for reasoning layer | Keep loosely coupled; ML.NET handles deterministic scoring, LLM adds explanation |
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
| Both structured ML predictions AND natural language reasoning | **Hybrid**: ML.NET for predictions + LLM for reasoning layer | Keep loosely coupled; ML.NET handles reproducible scoring, LLM adds explanation |
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+
+**Critical rule:** Do NOT use an LLM for tasks that ML.NET handles well (classification on tabular data, regression, clustering). LLMs are slower, more expensive, and non-deterministic for these tasks.
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md:35]:
```suggestion
**Critical rule:** Do NOT use an LLM for tasks that ML.NET handles well (classification on tabular data, regression, clustering). LLMs are slower, more expensive, and less predictable for these tasks.
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+- [ ] Non-deterministic outputs have validation and fallback paths
+- [ ] `dotnet build -c Release -warnaserror` completes cleanly
+
+## Anti-Patterns to Reject
+
+When reviewing or generating code, flag and redirect the developer if any of these patterns are detected:
+
+| Anti-pattern | Redirect |
+|-------------|----------|
+| Using an LLM for classification on structured/tabular data | Use ML.NET instead — it is faster, cheaper, and deterministic |
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md:327]:
```suggestion
| Using an LLM for classification on structured/tabular data | Use ML.NET instead — it is faster, cheaper, and reproducible given a fixed seed and dataset |
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+<PackageReference Include="Microsoft.ML.AutoML" Version="0.*" />
+<!-- Only if custom numerical work is needed: -->
+<PackageReference Include="MathNet.Numerics" Version="5.*" />
+<!-- Only for data exploration: -->
+<PackageReference Include="Microsoft.Data.Analysis" Version="0.*" />
+```
+
+> **Do NOT use** Accord.NET — it is archived and unmaintained.
+
+#### Modern AI packages
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md:85]:
```suggestion
#### Generative AI and Retrieval Augmented Generation (RAG) packages
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+
+Install only the packages needed for the selected technology branch. Do not mix competing abstractions.
+
+#### Classic ML packages
+
+```xml
+<PackageReference Include="Microsoft.ML" Version="4.*" />
+<PackageReference Include="Microsoft.ML.AutoML" Version="0.*" />
+<!-- Only if custom numerical work is needed: -->
+<PackageReference Include="MathNet.Numerics" Version="5.*" />
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md:78]:
```suggestion
PackageReference Include="System.Numerics.Tensors" Version="10.*"
<PackageReference Include="MathNet.Numerics" Version="5.*" />
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+<PackageReference Include="Microsoft.Extensions.AI" Version="9.*" />
+
+<!-- Orchestration (agents, workflows, tools, memory) — prerelease; use dotnet add package Microsoft.Agents.AI --prerelease -->
+<PackageReference Include="Microsoft.Agents.AI" Version="1.*-*" />
+
+<!-- Cloud LLM provider (pick one) -->
+<PackageReference Include="Azure.AI.OpenAI" Version="2.*" />
+<!-- OR -->
+<PackageReference Include="OpenAI" Version="2.*" />
+
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md:98]:
```suggestion

<!-- Client-side token counting for cost management -->
    <PackageReference Include="Microsoft.ML.Tokenizers" Version="2.*" 

```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+   ```
+
+3. **Retry logic**: Always implement retry with exponential backoff:
+   ```csharp
+   services.AddChatClient(builder => builder
+       .UseOpenAIChatClient(modelId)
+       .Use(new RetryingChatClient(maxRetries: 3)));
+   ```
+
+4. **Cost control**: Always estimate and log token usage. Choose the smallest model tier that meets quality requirements (e.g., `gpt-4o-mini` before `gpt-4o`).
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
4. **Cost control**: Always estimate and log token usage. Use Microsoft.ML.Tokenizers to count tokens client-side before sending requests so you can enforce budgets proactively. Choose the smallest model tier that meets quality requirements (e.g., gpt-4o-mini before gpt-4o).
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+
+1. **Iteration limits**: Always cap agentic loops to prevent runaway execution:
+   ```csharp
+   var settings = new AgentInvokeOptions
+   {
+       MaximumIterations = 10,
+   };
+   ```
+
+2. **Cost ceiling**: Implement a token budget per execution and terminate when reached.
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
2. **Cost ceiling**: Implement a token budget per execution and terminate when reached. Use Microsoft.ML.Tokenizers to count prompt and completion tokens locally and compare against the budget before each iteration.
```

File: plugins/dotnet-ai/skills/technology-selection/SKILL.md [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+| Using `Microsoft.SemanticKernel` for new projects | Use `Microsoft.Extensions.AI` + `Microsoft.Agents.AI` — Semantic Kernel is superseded by these newer abstractions for LLM orchestration and tool calling |
+
+## Common Pitfalls
+
+| Pitfall | Solution |
+|---------|----------|
+| Over-engineering with LLMs | Start with the simplest approach (rules, ML.NET) and add LLM capability only when simpler methods fall short |
+| Evaluating ML models on training data | Always use `TrainTestSplit` and report metrics on the held-out test set |
+| LLM output drift between deployments | Pin to specific dated model versions (e.g., `gpt-4o-2024-08-06`) |
+| Token cost surprises | Set `MaxOutputTokens`, log token counts per request, and alert on budget thresholds |
```
Comment (luisquintanilla) [inline plugins/dotnet-ai/skills/technology-selection/SKILL.md]:
```suggestion
| Token cost surprises | Set `MaxOutputTokens`, use Microsoft.ML.Tokenizers for accurate client-side token counting, log token counts per request, and alert on budget thresholds |
```
